Mathlib/Analysis/Calculus/EDeriv/Basic.lean

```diff
@@ -0,0 +1,176 @@
+and the second exterior derivative of a form is zero.
+-/
+
+open Filter ContinuousAlternatingMap Set
+open scoped Topology
+
+variable {𝕜 E F : Type*}
+  [NontriviallyNormedField 𝕜]
+  [NormedAddCommGroup E] [NormedSpace 𝕜 E]
+  [NormedAddCommGroup F] [NormedSpace 𝕜 F]
+  {n m k : ℕ} {r : WithTop ℕ∞}
+  {ω ω₁ ω₂ : E → E [⋀^Fin n]→L[𝕜] F} {s t : Set E} {x : E}
+
+/-- Exterior derivative of a differential form. -/
+noncomputable def ederiv (ω : E → E [⋀^Fin n]→L[𝕜] F) (x : E) : E [⋀^Fin (n + 1)]→L[𝕜] F :=
+  .uncurryFin (fderiv 𝕜 ω x)
+
+/-- Exterior derivative of a differential form within a set. -/
+noncomputable def ederivWithin (ω : E → E [⋀^Fin n]→L[𝕜] F) (s : Set E) (x : E) :
+    E [⋀^Fin (n + 1)]→L[𝕜] F :=
+  .uncurryFin (fderivWithin 𝕜 ω s x)
+
+@[simp]
+theorem ederivWithin_univ (ω : E → E [⋀^Fin n]→L[𝕜] F) :
+    ederivWithin ω univ = ederiv ω := by
+  ext1 x
+  rw[ederivWithin, ederiv, fderivWithin_univ]
+
+theorem ederivWithin_add (hsx : UniqueDiffWithinAt 𝕜 s x)
+    (hω₁ : DifferentiableWithinAt 𝕜 ω₁ s x) (hω₂ : DifferentiableWithinAt 𝕜 ω₂ s x) :
+    ederivWithin (ω₁ + ω₂) s x = ederivWithin ω₁ s x + ederivWithin ω₂ s x := by
+  simp [ederivWithin, fderivWithin_add hsx hω₁ hω₂, uncurryFin_add]
+
+theorem ederivWithin_fun_add (hsx : UniqueDiffWithinAt 𝕜 s x)
+    (hω₁ : DifferentiableWithinAt 𝕜 ω₁ s x) (hω₂ : DifferentiableWithinAt 𝕜 ω₂ s x) :
+    ederivWithin (fun x ↦ ω₁ x + ω₂ x) s x = ederivWithin ω₁ s x + ederivWithin ω₂ s x :=
+  ederivWithin_add hsx hω₁ hω₂
+
+theorem ederiv_add (hω₁ : DifferentiableAt 𝕜 ω₁ x) (hω₂ : DifferentiableAt 𝕜 ω₂ x) :
+    ederiv (ω₁ + ω₂) x = ederiv ω₁ x + ederiv ω₂ x := by
+  simp [← ederivWithin_univ, ederivWithin_add, *, DifferentiableAt.differentiableWithinAt]
+
+theorem ederiv_fun_add (hω₁ : DifferentiableAt 𝕜 ω₁ x) (hω₂ : DifferentiableAt 𝕜 ω₂ x) :
+    ederiv (fun x ↦ ω₁ x + ω₂ x) x = ederiv ω₁ x + ederiv ω₂ x :=
+  ederiv_add hω₁ hω₂
+
+theorem ederivWithin_smul (c : 𝕜) (ω : E → E [⋀^Fin n]→L[𝕜] F) (hsx : UniqueDiffWithinAt 𝕜 s x) :
+    ederivWithin (c • ω) s x = c • ederivWithin ω s x := by
+  simp [ederivWithin, fderivWithin_const_smul_of_field, hsx, uncurryFin_smul]
+
+theorem ederivWithin_fun_smul (c : 𝕜) (ω : E → E [⋀^Fin n]→L[𝕜] F)
+    (hsx : UniqueDiffWithinAt 𝕜 s x) :
+    ederivWithin (fun x ↦ c • ω x) s x = c • ederivWithin ω s x :=
+  ederivWithin_smul c ω hsx
+
+theorem ederiv_smul (c : 𝕜) (ω : E → E [⋀^Fin n]→L[𝕜] F) : ederiv (c • ω) x = c • ederiv ω x := by
+  simp [← ederivWithin_univ, ederivWithin_smul]
+
+theorem ederiv_fun_smul (c : 𝕜) (ω : E → E [⋀^Fin n]→L[𝕜] F) :
+    ederiv (c • ω) x = c • ederiv ω x :=
+  ederiv_smul c ω
+
+/-- Exterior derivative of a 0-form given by a function `f` within a set
+is the 1-form given by the derivative of `f` within the set. -/
+theorem ederivWithin_constOfIsEmpty (f : E → F) (hs : UniqueDiffWithinAt 𝕜 s x) :
+    ederivWithin (fun x ↦ constOfIsEmpty 𝕜 E (Fin 0) (f x)) s x =
```

Comment:
Can we have some more discoverable alias for 0-forms? `constOfIsEmpty` is not really teachable, alas.
Same for 1-forms.
Perhaps this can be a follow-up.

Reply:
I suggest that we leave this for a follow-up. I want to do sync `ContinuousMultilinearMap` API at the same time.